src/c3270/x3270if.man
'\" t
.TH X3270IF 1 "27 October 2008"
.SH "NAME"
x3270if \- command interface to x3270, c3270 and s3270
.SH "SYNOPSIS"
\fBx3270if\fP [option]... [ \fIaction\fP ]
.br
\fBx3270if \-i\fP
.SH "DESCRIPTION"
\fBx3270if\fP provides an interface between scripts and
the 3270 emulators \fIx3270\fP, \fIc3270\fP, and \fIs3270\fP.
.LP
\fBx3270if\fP operates in one of two modes.
In \fBaction mode\fP, it passes a single action and parameters to the
emulator for execution.
The result of the action is written to standard output, along with the
(optional) status of the emulator.
(The action is optional as well, so that \fBx3270if\fP can just reports the
emulator status.)
In \fBiterative mode\fP, it forms a continuous conduit between a script and
the emulator.
.LP
The \fIaction\fP takes the form:
.IP
\fIaction-name\fP(\fIparam\fP[,\fIparam\fP]...)
.LP
The parentheses are manatory, and usually must be quoted when \fBx3270if\fP is
called from a shell script.
.LP
If any \fIparam\fP contains a space or comma, it must be surrounded by
double quotes.
.SH "OPTIONS"
.TP
\fB\-s\fP \fIfield\fP
Causes \fBx3270if\fP to write to stdout the value of one of the
emulator status fields.
\fIField\fP is an integer in the range 0 through 11.
The value 0 is a no-op and is used only to return exit status indicating the
state of the emulator.
The indices 1-11 and meanings of each field are documented on the
\fIx3270-script\fP(1) manual page.
If an \fIaction\fP is specified as well, the status field is written after the
output of the action, separated by a newline.
The \fB\-s\fP option is mutually exclusive with the
\fB\-S\fP and \fB\-i\fP options.
.TP
\fB\-S\fP
Causes \fBx3270if\fP to write to stdout the value of all of the
emulator status fields.
If an \fIaction\fP is specified as well, the status fields are written after the
output of the action, separated by a newline.
The \fB\-S\fP option is mutually exclusive with the
\fB\-s\fP and \fB\-i\fP options.
.TP
\fB\-i\fP
Puts \fBx3270if\fP in iterative mode.
Data from \fBx3270if\fP's standard input is copied to the
emulator's script input; data from the
emulator's script output is copied to
\fBx3270if\fP's standard output.
The \fB\-i\fP option is mutually exclusive with the
\fB\-s\fP and \fB\-S\fP options.
\fBx3270if\fP
runs until it detects end-of-file on its standard input or on the
output from the emulator.
(This mode exists primarily to give \fIexpect\fP(1)
a process to run, on systems which do not support bidirectional pipes.)
.TP
\fB\-p\fP \fIprocess-id\fP
Causes \fIx3270if\fP to use a Unix-domain socket to connect to the emulator,
rather than pipe file descriptors given in environment variables.
The emulator must have been started with the \fB\-socket\fP option.
.TP
\fB\-v\fP
Turns on verbose debug messages, showing on stderr the literal data that is
passed between the emulator and \fBx3270if\fP.
.SH "EXIT STATUS"
In action mode, if the requested \fIaction\fP succeeds,
\fBx3270if\fP exits with status 0.
If the action fails, \fBx3270if\fP exits with status 1.
In iterative mode, \fBx3270if\fP
exits with status 0 when it encounters end-of-file.
If there is an operational error within \fBx3270if\fP
itself, such as a command-line syntax error, missing environment
variable, or an unexpectedly closed pipe,
\fBx3270if\fP exits with status 2.
.SH "ENVIRONMENT"
Unless the \fB\-socket\fP option is given when they are started,
\fIx3270\fP, \fIc3270\fP, and \fIs3270\fP
use a pair of pipes for communication with each child script process.
The values of the file descriptors for these pipes are encoded as text
in two environment variables, which are required by
\fBx3270if\fP:
.TP
\fBX3270OUTPUT\fP
Output from the emulator, input to the child process.
.TP
\fBX3270INPUT\fP
Input to the emulator, output from the child process.
.SH "SEE ALSO"
x3270(1),
c3270(1),
s3270(1),
x3270-script(1)
.SH "COPYRIGHT"
Copyright 1999, 2000, 2001, 2004, 2005 by Paul Mattes.
.RS
Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation.
.RE


src/x3270/x3270if.c
/*
 * Copyright 1995, 1996, 1999, 2000, 2001, 2002, 2004 by Paul Mattes.
 *  Permission to use, copy, modify, and distribute this software and its
 *  documentation for any purpose and without fee is hereby granted,
 *  provided that the above copyright notice appear in all copies and that
 *  both that copyright notice and this permission notice appear in
 *  supporting documentation.
 *
 * x3270, c3270, and s3270 are distributed in the hope that they will
 * be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file LICENSE
 * for more details.
 */

/*
 * Script interface utility for x3270, c3270 and s3270.
 *
 * Accesses an x3270 command stream on the file descriptors defined by the
 * environment variables X3270OUTPUT (output from x3270, input to script) and
 * X3270INPUT (input to x3270, output from script).
 */

#include "conf.h"
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <errno.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#if defined(HAVE_SYS_SELECT_H) /*[*/
#include <sys/select.h>
#endif /*]*/
#if defined(HAVE_GETOPT_H) /*[*/
#include <getopt.h>
#endif /*]*/

#define IBS	4096

#define NO_STATUS	(-1)
#define ALL_FIELDS	(-2)

extern int optind;
extern char *optarg;

static char *me;
static int verbose = 0;
static char buf[IBS];

static void iterative_io(int pid);
static void single_io(int pid, int fn, char *cmd);

static void
usage(void)
{
	(void) fprintf(stderr, "\
usage: %s [-v] [-S] [-s field] [-p pid] [action[(param[,...])]]\n\
       %s -i\n", me, me);
	exit(2);
}

/* Get a file descriptor from the environment. */
static int
fd_env(const char *name)
{
	char *fdname;
	int fd;

	fdname = getenv(name);
	if (fdname == (char *)NULL) {
		(void) fprintf(stderr, "%s: %s not set in the environment\n",
				me, name);
		exit(2);
	}
	fd = atoi(fdname);
	if (fd <= 0) {
		(void) fprintf(stderr, "%s: invalid value '%s' for %s\n", me,
		    fdname, name);
		exit(2);
	}
	return fd;
}

int
main(int argc, char *argv[])
{
	int c;
	int fn = NO_STATUS;
	char *ptr;
	int iterative = 0;
	int pid = 0;

	/* Identify yourself. */
	if ((me = strrchr(argv[0], '/')) != (char *)NULL)
		me++;
	else
		me = argv[0];

	/* Parse options. */
	while ((c = getopt(argc, argv, "ip:s:Sv")) != -1) {
		switch (c) {
		    case 'i':
			if (fn >= 0)
				usage();
			iterative++;
			break;
		    case 'p':
			pid = (int)strtoul(optarg, &ptr, 0);
			if (ptr == optarg || *ptr != '\0' || pid <= 0) {
				(void) fprintf(stderr,
				    "%s: Invalid process ID: '%s'\n", me,
				    optarg);
				usage();
			}
			break;
		    case 's':
			if (fn >= 0 || iterative)
				usage();
			fn = (int)strtol(optarg, &ptr, 0);
			if (ptr == optarg || *ptr != '\0' || fn < 0) {
				(void) fprintf(stderr,
				    "%s: Invalid field number: '%s'\n", me,
				    optarg);
				usage();
			}
			break;
		    case 'S':
			if (fn >= 0 || iterative)
				usage();
			fn = ALL_FIELDS;
			break;
		    case 'v':
			verbose++;
			break;
		    default:
			usage();
			break;
		}
	}

	/* Validate positional arguments. */
	if (optind == argc) {
		/* No positional arguments. */
		if (fn == NO_STATUS && !iterative)
			usage();
	} else {
		/* Got positional arguments. */
		if (iterative)
			usage();
	}

	/* Ignore broken pipes. */
	(void) signal(SIGPIPE, SIG_IGN);

	/* Do the I/O. */
	if (!iterative) {
		single_io(pid, fn, argv[optind]);
	} else {
		iterative_io(pid);
	}
	return 0;
}

/* Connect to a Unix-domain socket. */
static int
usock(int pid)
{
	struct sockaddr_un ssun;
	int fd;

	fd = socket(AF_UNIX, SOCK_STREAM, 0);
	if (fd < 0) {
		perror("socket");
		exit(2);
	}
	(void) memset(&ssun, '\0', sizeof(struct sockaddr_un));
	ssun.sun_family = AF_UNIX;
	(void) sprintf(ssun.sun_path, "/tmp/x3sck.%d", pid);
	if (connect(fd, (struct sockaddr *)&ssun, sizeof(ssun)) < 0) {
		perror("connect");
		exit(2);
	}
	return fd;
}

/* Do a single command, and interpret the results. */
static void
single_io(int pid, int fn, char *cmd)
{
	int sockfd;
	FILE *inf = NULL, *outf = NULL;
	char status[IBS] = "";
	int xs = -1;

	/* Verify the environment and open files. */
	if (pid) {
		sockfd = usock(pid);
		inf = fdopen(sockfd, "r");
		outf = fdopen(dup(sockfd), "w");
	} else {
		inf = fdopen(fd_env("X3270OUTPUT"), "r");
		outf = fdopen(fd_env("X3270INPUT"), "w");
	}
	if (inf == (FILE *)NULL) {
		perror("x3270if: input: fdopen");
		exit(2);
	}
	if (outf == (FILE *)NULL) {
		perror("x3270if: output: fdopen");
		exit(2);
	}

	/* Speak to x3270. */
	if (fprintf(outf, "%s\n", (cmd != NULL)? cmd: "") < 0 ||
	    fflush(outf) < 0) {
		perror("x3270if: printf");
		exit(2);
	}
	if (verbose)
		(void) fprintf(stderr, "i+ out %s\n",
		    (cmd != NULL) ? cmd : "");

	/* Get the answer. */
	while (fgets(buf, IBS, inf) != (char *)NULL) {
		int sl = strlen(buf);

		if (sl > 0 && buf[sl-1] == '\n')
			buf[--sl] = '\0';
		if (verbose)
			(void) fprintf(stderr, "i+ in %s\n", buf);
		if (!strcmp(buf, "ok")) {
			(void) fflush(stdout);
			xs = 0;
			break;
		} else if (!strcmp(buf, "error")) {
			(void) fflush(stdout);
			xs = 1;
			break;
		} else if (!strncmp(buf, "data: ", 6)) {
			if (printf("%s\n", buf+6) < 0) {
				perror("x3270if: printf");
				exit(2);
			}
		} else
			(void) strcpy(status, buf);
	}

	/* If fgets() failed, so should we. */
	if (xs == -1) {
		if (feof(inf))
			(void) fprintf(stderr,
				    "x3270if: input: unexpected EOF\n");
		else
			perror("x3270if: input");
		exit(2);
	}

	if (fflush(stdout) < 0) {
		perror("x3270if: fflush");
		exit(2);
	}

	/* Print status, if that's what they want. */
	if (fn != NO_STATUS) {
		char *sf = (char *)NULL;
		char *sb = status;
		int rc;

		if (fn == ALL_FIELDS) {
			rc = printf("%s\n", status);
		} else {
			do {
				if (!fn--)
					break;
				sf = strtok(sb, " \t");
				sb = (char *)NULL;
			} while (sf != (char *)NULL);
			rc = printf("%s\n", (sf != (char *)NULL) ? sf : "");
		}
		if (rc < 0) {
			perror("x3270if: printf");
			exit(2);
		}
	}

	if (fflush(stdout) < 0) {
		perror("x3270if: fflush");
		exit(2);
	}

	exit(xs);
}

/* Act as a passive pipe between 'expect' and x3270. */
static void
iterative_io(int pid)
{
#	define N_IO 2
	struct {
		const char *name;
		int rfd, wfd;
		char buf[IBS];
		int offset, count;
	} io[N_IO];	/* [0] is program->x3270, [1] is x3270->program */
	fd_set rfds, wfds;
	int fd_max = 0;
	int i;

#ifdef DEBUG
	if (verbose) {
		freopen("/tmp/x3270if.dbg", "w", stderr);
		setlinebuf(stderr);
	}
#endif

	/* Get the x3270 file descriptors. */
	io[0].name = "program->x3270";
	io[0].rfd = fileno(stdin);
	if (pid)
		io[0].wfd = usock(pid);
	else
		io[0].wfd = fd_env("X3270INPUT");
	io[1].name = "x3270->program";
	if (pid)
		io[1].rfd = dup(io[0].wfd);
	else
		io[1].rfd = fd_env("X3270OUTPUT");
	io[1].wfd = fileno(stdout);
	for (i = 0; i < N_IO; i++) {
		if (io[i].rfd > fd_max)
			fd_max = io[i].rfd;
		if (io[i].wfd > fd_max)
			fd_max = io[i].wfd;
		(void) fcntl(io[i].rfd, F_SETFL,
		    fcntl(io[i].rfd, F_GETFL, 0) | O_NDELAY);
		(void) fcntl(io[i].wfd, F_SETFL,
		    fcntl(io[i].wfd, F_GETFL, 0) | O_NDELAY);
		io[i].offset = 0;
		io[i].count = 0;
	}
	fd_max++;

	for (;;) {
		int rv;

		FD_ZERO(&rfds);
		FD_ZERO(&wfds);

		for (i = 0; i < N_IO; i++) {
			if (io[i].count) {
				FD_SET(io[i].wfd, &wfds);
#ifdef DEBUG
				if (verbose)
					(void) fprintf(stderr,
					    "enabling output %s %d\n",
					    io[i].name, io[i].wfd);
#endif
			} else {
				FD_SET(io[i].rfd, &rfds);
#ifdef DEBUG
				if (verbose)
					(void) fprintf(stderr,
					    "enabling input %s %d\n",
					    io[i].name, io[i].rfd);
#endif
			}
		}

		if ((rv = select(fd_max, &rfds, &wfds, (fd_set *)NULL,
				(struct timeval *)NULL)) < 0) {
			perror("x3270if: select");
			exit(2);
		}
		if (verbose)
			(void) fprintf(stderr, "select->%d\n", rv);

		for (i = 0; i < N_IO; i++) {
			if (io[i].count) {
				if (FD_ISSET(io[i].wfd, &wfds)) {
					rv = write(io[i].wfd,
					    io[i].buf + io[i].offset,
					    io[i].count);
					if (rv < 0) {
						(void) fprintf(stderr,
						    "x3270if: write(%s): %s",
						    io[i].name,
						    strerror(errno));
						exit(2);
					}
					io[i].offset += rv;
					io[i].count -= rv;
#ifdef DEBUG
					if (verbose)
						(void) fprintf(stderr,
						    "write(%s)->%d\n",
						    io[i].name, rv);
#endif
				}
			} else if (FD_ISSET(io[i].rfd, &rfds)) {
				rv = read(io[i].rfd, io[i].buf, IBS);
				if (rv < 0) {
					(void) fprintf(stderr,
					    "x3270if: read(%s): %s",
					    io[i].name, strerror(errno));
					exit(2);
				}
				if (rv == 0)
					exit(0);
				io[i].offset = 0;
				io[i].count = rv;
#ifdef DEBUG
				if (verbose)
					(void) fprintf(stderr,
					    "read(%s)->%d\n", io[i].name, rv);
#endif
			}
		}
	}
}
